rdt/transformers/base.py

```diff
@@ -1,55 +1,229 @@
+        """Return the suggested next transformer to be used for each column.
+
+        Returns:
+            dict:
+                Mapping from transformed column names to the transformers to apply to each column.
+        """
+        return self._add_prefix(self.NEXT_TRANSFORMERS)
+
+    def fit(self, data, columns):
+        """Fit the transformer to the `columns` of the `data`.
+
+        Args:
+            data (pandas.DataFrame):
+                The entire table.
+            columns (list):
+                Column names. Must be present in the data.
+        """
+        # make sure columns is a list where every column is in the data
+        if isinstance(columns, tuple) and columns not in data:
+            columns = list(columns)
+        elif not isinstance(columns, list):
+            columns = [columns]
+        if any(column not in data for column in columns):
+            raise KeyError(f'Some of the columns were not present in the data.')
+
+        self._column_prefix = '#'.join(columns)
+        self._output_columns = list(self.get_output_types().keys())
+
+        # make sure none of the generated `output_columns` exists in the data
+        while any(output_column in data for output_column in self._output_columns):
+            self._column_prefix += '#'
+            self._output_columns = list(self.get_output_types().keys())
+
+        self._columns = columns
+        columns = self._convert_if_length_one(self._columns)
+        self._fit(data[columns])
+
+    def _convert_if_length_one(self, columns):
+        """Convert columns to string if it's a list of length one."""
+        if len(columns) == 1:
+            columns = columns[0]
+
+        return columns
+
+    def _fit(self, columns_data):
         """Fit the transformer to the data.
 
         Args:
-            data (pandas.Series or numpy.array):
+            columns_data (pandas.DataFrame or pandas.Series):
                 Data to transform.
         """
         raise NotImplementedError()
 
     def transform(self, data):
+        """Transform the `self._columns` of the `data`.
+
+        Args:
+            data (pandas.DataFrame):
+                The entire table.
+
+        Returns:
+            pd.DataFrame:
+                The entire table, containing the transformed data.
+        """
+        # if `data` doesn't have the columns that were fitted on, don't transform
+        if any(column not in data.columns for column in self._columns):
+            return data
+
+        data = data.copy()
+
+        columns = self._convert_if_length_one(self._columns)
+        columns_data = data[columns]
+        transformed_data = self._transform(columns_data)
+
+        output_columns = self._convert_if_series(self._output_columns, transformed_data)
+        data[output_columns] = transformed_data
+        data.drop(self._columns, axis=1, inplace=True)
+
+        return data
+
+    def _convert_if_series(self, columns, data):
```

Comment:
can you move this before `transform` since `transform` uses it?

Reply:
I prefer to write the helpers after the function that calls them. Do you know if writing them the other way around is the general guideline? Because then I'd have to change the order of all three helpers as well.

Reply:
As far as I am concerned, everywhere in SDV lists the helpers before the methods that use them.

Reply:
Huh, I hadn't noticed. I guess I can invert them then 👍